File: src/main/java/net/challenge/client/injection/mixins/gui/MixinGuiInGame.java
/*
 * Challenge Client
 * https://github.com/MC-Challenge/ChallengeClient/
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.challenge.client.injection.mixins.gui;

import net.challenge.client.core.ClientCore;
import net.challenge.client.events.Render2DEvent;
import net.minecraft.client.gui.GuiIngame;
import net.minecraft.client.gui.ScaledResolution;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

@Mixin(GuiIngame.class)
@SideOnly(Side.CLIENT)
public class MixinGuiInGame {

    @Inject(method = "renderTooltip", at = @At("RETURN"))
    public void renderTooltip(ScaledResolution sr, float partialTicks, CallbackInfo callbackInfo) {
        ClientCore.INSTANCE.getEventBus().post(new Render2DEvent(sr, partialTicks));
    }
}

File: src/main/kotlin/net/challenge/client/injection/mixins/gui/MixinGuiInGame.kt
package net.challenge.client.injection.mixins.gui

import net.challenge.client.core.ClientCore
import net.challenge.client.events.Render2DEvent
import net.minecraft.client.gui.GuiIngame
import net.minecraft.client.gui.ScaledResolution
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo

@Mixin(GuiIngame::class)
@SideOnly(Side.CLIENT)
class MixinGuiInGame {

    // TODO is the injection correct? TEST!
    @Inject(method = "renderTooltip", at = [(At("HEAD"))])
    private fun renderTooltip(sr: ScaledResolution, partialTicks: Float, callbackInfo: CallbackInfo) {
        ClientCore.eventBus.post(Render2DEvent(sr, partialTicks))
    }
}